src/typecheck/arith.ml

```diff
@@ -11,32 +11,69 @@ module Ae = struct
       (Type : Tff_intf.S)
       (Ty : Dolmen.Intf.Ty.Ae_Arith with type t := Type.Ty.t)
       (T : Dolmen.Intf.Term.Ae_Arith with type t := Type.T.t
-                                        and type ty := Type.Ty.t) = struct
+                                      and type ty := Type.Ty.t) = struct
 
-      type _ Type.err +=
+    type _ Type.err +=
       | Expected_arith_type : Type.Ty.t -> Term.t Type.err
 
-      let dispatch1 env (mk_int, mk_real) ast t =
-        let ty = T.ty t in
-        match Ty.view ty with
-        | `Int -> mk_int t
-        | `Real -> mk_real t
-        | _ -> Type._error env (Ast ast) (Expected_arith_type ty)
-
-      let dispatch2 env (mk_int, mk_real) ast a b =
-        let tya = T.ty a in
-        match Ty.view tya with
-        | `Int -> mk_int a b
-        | `Real -> mk_real a b
-        | _ -> Type._error env (Ast ast) (Expected_arith_type tya)
-
-      let promote_to_real t =
-        let ty = T.ty t in
-        match Ty.view ty with
-        | `Int -> T.Int.to_real t
-        | `Real -> t
-        (* this will result in a more precise typing error later, so it's okay *)
-        | _ -> t
+    let dispatch1 env (mk_int, mk_real) ast t =
+      let ty = T.ty t in
+      match Ty.view ty with
+      | `Int -> mk_int t
+      | `Real -> mk_real t
+      | _ -> Type._error env (Ast ast) (Expected_arith_type ty)
+
+    let dispatch2 env (mk_int, mk_real) ast a b =
+      let tya = T.ty a in
+      match Ty.view tya with
+      | `Int -> mk_int a b
+      | `Real -> mk_real a b
+      | _ -> Type._error env (Ast ast) (Expected_arith_type tya)
+
+    let promote_to_real t =
+      let ty = T.ty t in
+      match Ty.view ty with
+      | `Int -> T.Int.to_real t
+      | `Real -> t
+      (* this will result in a more precise typing error later, so it's okay *)
+      | _ -> t
+
+    let parse_bound env ast =
+      match ast.Term.term with
+      | Symbol { ns = Term; name = Simple "?" } -> None
+      | _ ->
+        let t = Type.parse_term env ast in
+        Some t
+
+    let parse_in_interval env ~strict_lower ~strict_upper _ (ast, lower, upper) =
+      let t = Type.parse_term env ast in
+      let t_ty = T.ty t in
+      let x = Type.T.Var.mk "x" t_ty in
+      let x_t = Type.T.of_var x in
```

Comment:
Is this not guaranteed to be `t_ty`?

Reply:
Ah no, it's a bit obscured by all the functors, but in this case, `Type` is actually the typechecker, and therefore `Type.T` is the term module that was given to instantiate the typechecker functor, so `x_t` is actually the term (and not type) that simply represents the variable `x`.